broker/nats/nats.go

```diff
@@ -207,39 +197,59 @@ func NewBroker(opts ...broker.Option) broker.Broker {
 		Context: context.Background(),
 	}
 
-	for _, o := range opts {
-		o(&options)
+	n := &natsBroker{
+		opts: options,
 	}
+	n.setOption(opts...)
 
-	natsOpts := nats.GetDefaultOptions()
-	if n, ok := options.Context.Value(optionsKey{}).(nats.Options); ok {
-		natsOpts = n
+	return n
+}
+
+func (n *natsBroker) setOption(opts ...broker.Option) {
+	for _, o := range opts {
+		o(&n.opts)
 	}
 
-	var drain bool
-	if _, ok := options.Context.Value(drainSubscriptionKey{}).(bool); ok {
-		drain = true
+	n.Once.Do(func() {
```

Comment:
why you guard this? setOption called in Init and in NewBroker

Reply:
This is because we need only set default option in the first time. Otherwise, the nats option set in `New` will be reset in `Init`

Reply:
but New and Init both have opts and needs to allow overwrite options stored in broker.... @asim i'm right?

Reply:
Yea, Init is for reinitialising options

Reply:
But I guess what this code is saying is that we only want to fetch default options once. We don't want to overwrite existing options with default options. So it sort of makes sense.

Reply:
e.g New sets default options and adds addrs then Init adds something else, you don't want init to replace addrs

Reply:
> But I guess what this code is saying is that we only want to fetch default options once. We don't want to overwrite existing options with default options. So it sort of makes sense.

makes sense.

Reply:
Asim is right, it is what i mean here. Though I can kind of add some code seperately in `New()` only, instead of using `sync.Once`. While i think it seems more explicit to put them together.